docs/modules/SDDs/pages/0033-commodore-component-instance-versioning.adoc

```diff
@@ -0,0 +1,91 @@
+In a valid catalog, all used versions of the same component must advertise this support.
+
+Components with multi-versioning support must ensure they create no file conflicts when downloading dependencies.
+In practice, this means components must use the parameter `${_base_directory}` in their compile configuration (`parameters.kapitan.dependencies` and `parameters.kapitan.compile`).
+
+Support for providing different versions of Jsonnet dependencies is currently not implemented.
+Jsonnet dependencies are always provided from the main (non-instantiated) version of the component.
+If a component is to support multi-versioning, Jsonnet dependencies should therefore be compatible within a reasonable span of versions.
+
+Similarly, if the component includes a Jsonnet library that is made available to other components, the version for the library is taken from the main (non-instantiated) version of the component.
+
+=== Implementation Details/Notes/Constraints
+
+We can clone the repository of the base component and create a separate worktree for each instance to check out the correct version.
+
+If the instance version specifies a different URL from the base component, more work is needed: in this case, we clone the changed repository for the instance and create a worktree from that.
+
+This means that if two instances both override the URL to the same value, this repository is cloned twice, which can impact performance.
+However, since this only affects instances which override the repository URL, the impact is tolerable.
+
+=== Risks and Drawbacks
+
+* Since the JSonnet dependencies and libraries are all provided from a single component version, there is a risk for some combinations of versions not compiling properly due to version compatibility issues.
+** Even if a component's dependencies are sufficiently backwards-compatible at the moment, it's possible for future changes to break compatibility.
+This might force a component author to remove multi-version support in higher component versions, causing a catalog to stop compiling if either component instance is updated.
+* Overriding the repository URL in a component instance always causes a new repository to be cloned - even if the same URL is used in multiple instances.
+
+== Alternatives
+
+=== Automatically detecting multi-version support
+
+It could be possible to automatically determine whether a component supports multi-versioning, by way of checking whether the `${_base_directory}` parameter is used in the component's compile configuration.
+However, automatically declaring components to be "safe for multi-versioning" carries some risk - even if a component is configured accordingly, it is not guaranteed that its provided JSonnet libraries and/or its JSonnet dependencies work if used across versions.
+Therefore, we deem it safer to let component authors decide if multi-versioning is safe to use for their component.
```

Comment:
This conclusion is based on my vague understanding of how stable these dependencies typically are, so happy to discuss.

Reply:
```suggestion
Additionally, so far we've defined explicit flags for components to indicate that they support optional features, such as multiple instances.
For all of those reasons, we've decided that component authors must explicitly declare whether multi-versioning is safe to use for their component.
```